plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase       #-}
+module Language.PlutusIR.Transform.LetFloat (floatTerm) where
+
+import           Language.PlutusIR
+import           Language.PlutusIR.Analysis.Dependencies
+import           Language.PlutusIR.Analysis.Free
+
+import           Control.Lens
+import           Control.Monad.Reader
+import           Control.Monad.RWS
+import           Control.Monad.State
+
+import qualified Language.PlutusCore                     as PLC
+import qualified Language.PlutusCore.Name                as PLC
+
+import qualified Algebra.Graph.AdjacencyMap              as AM
+import qualified Algebra.Graph.AdjacencyMap.Algorithm    as AM
+import qualified Algebra.Graph.NonEmpty.AdjacencyMap     as AMN
+
+import           Data.Foldable                           (fold)
+import           Data.Function                           (on)
+import qualified Data.IntMap                             as IM
+import qualified Data.Map                                as M
+import           Data.Maybe                              (fromMaybe)
+import qualified Data.Set                                as S
+
+{- Note: [Float algorithm]
+
+Our algorithm is influenced by the algorithm described in Section-4 of:
+Peyton Jones, Simon, Will Partain, and Andre Santos. "Let-Floating: Moving Bindings to Give Faster Programs." In Proceedings of the First ACM SIGPLAN International Conference on Functional Programming, 1-12. ICFP '96. New York, NY, USA: ACM, 1996. https://doi.org/10.1145/232627.232630.
+
+Our algorithm tries to do two things:
+
+a) move let-bindings as outwards as possible (a.k.a. full laziness).
```

Comment:
Full laziness is dangerous even in a lazy language (because a binding may live more than it's supposed to), but in our strict setting it doesn't sound appealing at all. So let's say we have a Plutus equivalent of

```haskell
if b
    then
        let x1 = <expensive_computation_1> in
            y1
    else
        let x2 = <expensive_computation_2> in
            y2
```

will it be transformed into

```haskell
let x1 = <expensive_computation_1>
     x2 = <expensive_computation_2>
in if b then y1 else y2
```

? It'd be fine in Haskell, but we certainly don't want this in Plutus. 

So I'd like to see a discussion on how exactly things are intended to work. A few examples would also help.